hal/src/dma/controller.rs

```diff
@@ -0,0 +1,186 @@
+        ///
+        /// This is a newtype which impls the `hal::dma::controller::Registers` trait.
+        pub struct $newtype;
+
+        impl core::ops::Deref for $newtype {
+            type Target = $target;
+
+            fn deref(&self) -> &$target {
+                *$inner
+            }
+        }
+
+        impl_dmac_registers!($newtype);
+    }
+}
+
+/// SAMD21 DMAC registers
+#[cfg(not(feature = "samd51"))]
+mod samd21 {
+    /// Number of DMA channels
+    pub const NUM_CHANNELS: usize = 1;
+
+    /// DMAC registers (SAMD21)
+    pub trait Registers {
+        // Define accessors for registers common to SAMD21 and SAMD51
+        dmac_common_registers!();
+        dmac_io_register!(read_qosctrl, write_qosctrl, u8, "0x0e - QOS control");
+        dmac_io_register!(read_chid, write_chid, u8, "0x3f - Channel ID");
+        dmac_io_register!(read_chctrla, write_chctrla, u8, "0x40 - Channel control (A)");
+        dmac_io_register!(read_chctrlb, write_chctrlb, u32, "0x44 - Channel control (B)");
+        dmac_io_register!(read_chintenclr, write_chintenclr, u8, "0x4c - Channel interrupt enable clear");
+        dmac_io_register!(read_chintenset, write_chintenset, u8, "0x4d - Channel interrupt enable set");
+        dmac_io_register!(read_chintflag, write_chintflag, u8, "0x4e - Channel interrupt flag");
+        dmac_register!(read_chstatus, u8, "0x4f - Channel status");
+    }
+
+    #[macro_export]
+    macro_rules! impl_dmac_registers {
+        ($dmac:ty) => {
+            impl $crate::dma::controller::Registers for $dmac {
+                impl_dmac_common_registers!();
+                impl_dmac_io_register!(qosctrl, read_qosctrl, write_qosctrl, QOSCTRL, u8);
+                impl_dmac_io_register!(chid, read_chid, write_chid, CHID, u8);
+                impl_dmac_io_register!(chctrla, read_chctrla, write_chctrla, CHCTRLA, u8);
+                impl_dmac_io_register!(chctrlb, read_chctrlb, write_chctrlb, CHCTRLB, u32);
+                impl_dmac_io_register!(chintenclr, read_chintenclr, write_chintenclr, CHINTENCLR, u8);
+                impl_dmac_io_register!(chintenset, read_chintenset, write_chintenset, CHINTENSET, u8);
+                impl_dmac_io_register!(chintflag, read_chintflag, write_chintflag, CHINTFLAG, u8);
+                impl_dmac_register!(chstatus, read_chstatus, CHSTATUS, u8);
+            }
+        }
+    }
+}
+
+#[cfg(not(feature = "samd51"))]
+pub use samd21::*;
+
+/// SAMD51 DMAC registers
+#[cfg(feature = "samd51")]
+mod samd51 {
+    /// Number of DMA channels
+    pub const NUM_CHANNELS: usize = 32;
+
+    /// DMAC registers (SAMD51)
+    pub trait Registers {
+        // Define accessors for registers common to SAMD21 and SAMD51
+        dmac_common_registers!();
+    }
+
+    #[macro_export]
+    macro_rules! impl_dmac_registers {
+        ($dmac:ty) => {
+            impl $crate::dma::controller::Registers for $dmac {
+                impl_dmac_common_registers!();
```

Comment:
TODO: SAMD51 DMA channels. I figured I'd cross-check this basic approach before working on those.